core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -460,54 +459,51 @@ public Path getWorkingDirectory() {
   /**
    * {@inheritDoc}
    *
-   * Sets up a lazy connection to Alluxio through mFileSystem. This method will override and
-   * invalidate the current contexts. This must be called before client operations in order to
-   * guarantee the integrity of the contexts, meaning users should not alternate between using the
-   * Hadoop compatible API and native Alluxio API in the same process.
+   * Sets up a lazy connection to Alluxio through mFileSystem and mFsContext. This must be
+   * called before client operations in order to guarantee the integrity of the
+   * {@link FileSystemContext}.
    *
-   * If hadoop file system cache is enabled, this method should only be called when switching user.
+   * If hadoop file system cache is enabled, this method should only be called when switching
+   * user. If it is called twice on the same object an exception will be thrown.
    */
   @SuppressFBWarnings("ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD")
   @Override
   public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throws IOException {
     Preconditions.checkArgument(uri.getScheme().equals(getScheme()),
         PreconditionMessage.URI_SCHEME_MISMATCH.toString(), uri.getScheme(), getScheme());
-    if (mFsContext != null) {
-      throw new IOException("The AbstractFileSystem has already been initialized.");
-    }
-    super.initialize(uri, conf);
-    LOG.debug("initialize({}, {}). Connecting to Alluxio", uri, conf);
-    HadoopUtils.addSwiftCredentials(conf);
-    setConf(conf);
-
-    // HDFS doesn't allow the authority to be empty; it must be "/" instead.
-    String authority = uri.getAuthority() == null ? "/" : uri.getAuthority();
-    mAlluxioHeader = getScheme() + "://" + authority;
-    // Set the statistics member. Use mStatistics instead of the parent class's variable.
-    mStatistics = statistics;
-
-    Authority auth = Authority.fromString(uri.getAuthority());
-    if (auth instanceof UnknownAuthority) {
-      throw new IOException(String.format("Authority \"%s\" is unknown. The client can not be "
-              + "configured with this authority.", auth.toString()));
-    }
-
-    mUri = URI.create(mAlluxioHeader);
-
-    Map<String, Object> uriConfProperties = getConfigurationFromUri(uri);
-    AlluxioProperties alluxioProps = ConfigurationUtils.defaults();
-    Subject subject = getHadoopSubject();
-    if (subject != null) {
-      LOG.debug("Using Hadoop subject: {}", subject);
-    } else {
-      LOG.debug("No Hadoop subject. Using FileSystem Context without subject.");
-    }
-
     synchronized (this) {
```

Comment:
To minimize impact of the check on the rest of the method, we can do it like this:
```java
private final AtomicBoolean mCalledInitialize = new AtomicBoolean(false);
public void initialize() {
  Preconditions.checkState(mCalledInitialize.compareAndSet(false, true), "Illegally invoked initialize() multiple times");
  // normal implementation
```
}

```

Reply:
If initialize throws an exception is it expected then for the user to just create a new object?

Reply:
Users should never interact with an uninitialized `FileSystem` object. To create a filesystem, they should call `FileSystem.get(...)`, which will create the `FileSystem` and also call `initialize()`.

If users are advanced/unwise enough to directly instantiate `alluxio.hadoop.FileSystem` and they want to recover from exceptions thrown from `initialize`, I think it's fine for them to have to re-create the `FileSystem` object each time. I don't know why anyone would want to do this though when they can just use `FileSystem.get` or use our native API